src/sync.rs

```diff
@@ -0,0 +1,302 @@
+    }
+
+    #[inline]
+    /// Get the id of the thread that currently owns this lock.
+    fn mutex_get_owner(&mut self, id: MutexId) -> ThreadId {
+        let this = self.eval_context_ref();
+        this.machine.threads.sync.mutexes[id].owner.unwrap()
+    }
+
+    #[inline]
+    /// Check if locked.
+    fn mutex_is_locked(&mut self, id: MutexId) -> bool {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.mutexes[id].owner.is_some()
+    }
+
+    /// Lock by setting the mutex owner and increasing the lock count.
+    fn mutex_lock(&mut self, id: MutexId, thread: ThreadId) {
+        let this = self.eval_context_mut();
+        let mutex = &mut this.machine.threads.sync.mutexes[id];
+        if let Some(current_owner) = mutex.owner {
+            assert_eq!(thread, current_owner, "mutex already locked by another thread");
+            assert!(
+                mutex.lock_count > 0,
+                "invariant violation: lock_count == 0 iff the thread is unlocked"
+            );
+        } else {
+            mutex.owner = Some(thread);
+        }
+        mutex.lock_count = mutex.lock_count.checked_add(1).unwrap();
+    }
+
+    /// Unlock by decreasing the lock count. If the lock count reaches 0, unset
+    /// the owner.
+    fn mutex_unlock(&mut self, id: MutexId) -> Option<(ThreadId, usize)> {
+        let this = self.eval_context_mut();
+        let mutex = &mut this.machine.threads.sync.mutexes[id];
+        if let Some(current_owner) = mutex.owner {
+            mutex.lock_count = mutex
+                .lock_count
+                .checked_sub(1)
+                .expect("invariant violation: lock_count == 0 iff the thread is unlocked");
+            if mutex.lock_count == 0 {
+                mutex.owner = None;
+            }
+            Some((current_owner, mutex.lock_count))
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    /// Take a thread out the queue waiting for the lock.
+    fn mutex_enqueue(&mut self, id: MutexId, thread: ThreadId) {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.mutexes[id].queue.push_back(thread);
+    }
+
+    #[inline]
+    /// Take a thread out the queue waiting for the lock.
+    fn mutex_dequeue(&mut self, id: MutexId) -> Option<ThreadId> {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.mutexes[id].queue.pop_front()
+    }
+
+    #[inline]
+    /// Create state for a new read write lock.
+    fn rwlock_create(&mut self) -> RwLockId {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks.push(Default::default())
+    }
+
+    #[inline]
+    /// Check if locked.
+    fn rwlock_is_locked(&mut self, id: RwLockId) -> bool {
+        let this = self.eval_context_mut();
+        this.machine.threads.sync.rwlocks[id].writer.is_some()
+            || !this.machine.threads.sync.rwlocks[id].readers.is_empty()
```

Comment:
I think this is a good example for postfix `.not()` being clearer than prefix `!`.

Reply:
Thanks! I was not aware of this function.